_posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown

```diff
@@ -0,0 +1,192 @@
+
+## 4. Application code
+
+### a. Models
+
+- All dynamic finder methods except for `.find_by_...` are deprecated:
+
+```ruby
+# before:
+Authentication.find_all_by_provider_and_uid(provider, uid)
+
+# after:
+Authentication.where(provider: provider, uid: uid)
+```
+
+You can regain usage of these finders by adding the gem
+[activerecord-deprecated_finders](https://github.com/rails/activerecord-deprecated_finders)
+
+- ActiveRecord scopes now need a lambda:
+
+```ruby
+# before:
+default_scope where(deleted_at: nil)
+
+# after:
+default_scope { where(deleted_at: nil) }
+
+# before:
+has_many :posts, order: 'position'
+
+# after:
+has_many :posts, -> { order('position') }
+```
+
+- Protected Attributes are deprecated, but you can add the [gem for it](https://github.com/rails/protected_attributes).
+However, since the Rails core team dropped its support since Rails 5.0, you
+should begin migrating your models to Strong Parameters anyway.
+
+- ActiveRecord Observers were removed from the Rails 4.0 codebase and extracted
+into a gem. You can regain usage by adding the gem to your Gemfile:
+
+```ruby
+gem 'rails-observers' # https://github.com/rails/rails-observers
+```
+
+As an alternative, you can take a look at the [wisper gem](https://github.com/krisleech/wisper),
+or Rails' Concerns (which were added in Rails 4.0) for a slightly different
+approach.
+
+- ActiveResource was removed and extracted into its own gem:
+
+```ruby
+gem 'active_resource' # https://github.com/rails/activeresource
+```
+
+### b. Controllers
+
+- ActionController Sweeper was extracted into the `rails-observers` gem, you can
+regain usage by adding the gem to your Gemfile:
+
+```ruby
+gem 'rails-observers' # https://github.com/rails/rails-observers
+```
+
+- Action caching was extracted into its own gem, so if you're using this feature
+through either:
+
+```ruby
+caches_page   :public
+```
+
+or:
+
+```ruby
+caches_action :index, :show
+```
+
+You will need to add the gem:
+
+```ruby
+gem 'actionpack-action_caching' # https://github.com/rails/actionpack-action_caching
+```
+
+## 5. Tests
+
+From Ruby 1.9.x onwards, you have to include the [`test-unit` gem](https://rubygems.org/gems/test-unit)
+in your Gemfile as it was removed from the standard lib. As an alternative,
+migrate to `Minitest`, `RSpec` or your favorite test framework.
+
+## 6. Miscellaneous
+
+- Routes now require you to specify the request method, so you no longer can
+rely on the catch-all default.
+
+```ruby
+# change:
+match '/home' => 'home#index'
+
+# to:
+match '/home' => 'home#index', via: :get
+
+# or:
+get '/home' => 'home#index'
+```
+
+- Rails 4.0 dropped support for plugins, so you'll need to replace them with gems,
+either by searching for the project on [RubyGems](https://rubygems.org)/[Github](https://github.com),
+or by moving the plugin to your `lib` directory and require it from somewhere
+within your Rails app.
+
+## 7. Next steps
+
+If you're running Rails 4.0, congratulations!
```

Comment:
@mauro-oto This is kind of weird. When they read this step, supposedly they followed your steps and you can write something like this: 

```
If you successfully followed all these steps by now you're running Rails 4.0. So congratulations!
```